Large Send Offload for UDP protocol.

yazimut opened the pull request:
Added support for LSO for UDP protocol.

* Issues fixed (if any):
   ** Alignment in some functions.

yazimut committed:
LSO for UDP protocol support.

Added:
 * "--lso-udp SegSize" option

Fixed:
 * Alignment in function "iperf_parse_arguments".

yazimut committed:
Done Large Segmentation Offload for UDP protocol.

cail commented on src/iperf_api.c:
```diff
@@ -104,6 +105,8 @@
     #define UDP_SEGMENT 103 /* Set GSO segmentation size */
 #endif /* UDP_SEGMENT */
 
+#define LSO_UDP_MIN_SEGSIZE 1024
```
Why? thats not true! any size, including 1 byte works!

cail commented on src/iperf_udp.c:
```diff
@@ -567,6 +643,26 @@ iperf_udp_connect(struct iperf_test *test)
+
+        if (test->settings->domain == PF_UNSPEC ||
+            test->settings->domain == PF_INET) {
+                level = SOL_IP;
```
why need this?

cail commented on src/iperf_udp.c:
```diff
@@ -98,9 +106,11 @@ iperf_udp_recv(struct iperf_stream *sp)
 	    memcpy(&sec, sp->buffer, sizeof(sec));
 	    memcpy(&usec, sp->buffer+4, sizeof(usec));
 	    memcpy(&pc, sp->buffer+8, sizeof(pc));
+        memcpy(&lso_part, sp->buffer+12, sizeof(lso_part));
```
alignment. check other lines as well.

yazimut replied:
Fixed

cail replied:
but all other places??

cail commented on src/iperf_api.c:
```diff
@@ -1379,6 +1382,51 @@ iperf_parse_arguments(struct iperf_test *test, int argc, char **argv)
         warning("Debug output (-d) may interfere with JSON output (-J)");
     }
 
+    if (test->settings->lso_udp) {
```
not all linuxes support GSO_UDP socket option. We should somehow query that before use.

yazimut replied:
I've made a check for supporting (iperf_api.c - line 98). But it works only at compile time. In runtime, using of UDP LSO will cause the error in "sendmsg" call - EINVAL.

cail replied:
think that makes sense. EINVAL is enouth

yazimut committed:
Added runtime checking for support UDP LSO by configuration.
Removed bottom bound for UDP LSO Segment size.
Fixed some misspellings and alignments.

yazimut committed:
Fixed runtime checking

yazimut committed:
Now UDP LSO segment can be greater than 129 bytes.
Fixed empty lines in "iperf_api.c".

cail commented on docs/invoking.rst:
```diff
@@ -399,10 +399,17 @@ the executable.
+
+          --lso-udp n
+                 enables Large Segmentation Offload for UDP protocol. "n" is a
+                 segment size, it cannot be less 1024 bytes or greater than MTU.
```
remove notice on 1024b limit.

cail commented on src/iperf_api.c:
```diff
@@ -90,6 +95,16 @@
 #include <sys/sysinfo.h>
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 18, 0)
+    #error Your configuration isn't supporting LSO dor UDP! If you want to use it on\
```
thats wrong. this way you disable iperf builds on all linuxes < 4.18.
thats wrong.

cail commented on src/iperf_api.c:
```diff
@@ -804,25 +819,26 @@ iperf_parse_arguments(struct iperf_test *test, int argc, char **argv)
         {"nstreams", required_argument, NULL, OPT_NUMSTREAMS},
         {"xbind", required_argument, NULL, 'X'},
 #endif
-	{"pidfile", required_argument, NULL, 'I'},
```
you should not touch other code formatting. please check your editor (probably autoformat was applied).
such changes are unrelated to our pull request.

cail commented on src/iperf_udp.c:
```diff
@@ -87,20 +93,24 @@ iperf_udp_recv(struct iperf_stream *sp)
 	    sec = ntohl(sec);
 	    usec = ntohl(usec);
 	    pcount = ntohl(pc);
+        lso_part = ntohs(lso_part);
```
here

cail commented on src/iperf_udp.c:
```diff
@@ -214,31 +228,102 @@ iperf_udp_send(struct iperf_stream *sp)
 
 	uint32_t  sec, usec;
 	uint64_t  pcount;
+    uint16_t  lso_part = 0;
```
here. always review your pull request by yourself first...

cail commented on src/iperf_udp.c:
```diff
@@ -214,31 +228,102 @@ iperf_udp_send(struct iperf_stream *sp)
-	memcpy(sp->buffer, &sec, sizeof(sec));
-	memcpy(sp->buffer+4, &usec, sizeof(usec));
-	memcpy(sp->buffer+8, &pcount, sizeof(pcount));
+    lso_part = htons(lso_part);
```
here

cail commented on src/iperf_udp.c:
```diff
@@ -567,6 +652,34 @@ iperf_udp_connect(struct iperf_test *test)
+                val   = IPV6_PMTUDISC_DO;
+        }
+
+        if (setsockopt(s, level, name, &val, sizeof(val))) {
```
again, why you've added this? why you need MTU_DISCOVER for lso_udp specifically?
